File: 78. Subsets.java
class Solution {
    
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList();
        bk(nums, new ArrayList(), 0, result);
        return result;        
    }
    
    public void bk(int[] nums, List<Integer> tmp, int p, List<List<Integer>> result) {
        result.add(new ArrayList(tmp));
        if(p >= nums.length) {            
            return;
        }
        for(int i = p; i < nums.length; i++) {
            if( !tmp.contains(nums[i]) ) {
                tmp.add(nums[i]);
            }
            bk(nums, tmp, i+1, result);
            if(tmp.get(tmp.size() - 1) == nums[i]) {
                tmp.remove(tmp.size() - 1);
            }
        }
        
    }
}

File: 78. Subsets.kt
class Solution {
    val result = mutableListOf<List<Int>>()
    fun subsets(nums: IntArray): List<List<Int>> {
        backtracking(nums, 0, mutableListOf())
        return result
    }
    private fun backtracking(nums: IntArray, idx: Int, currentList: MutableList<Int>) {
        result.add(ArrayList(currentList))
        for (i in idx until nums.size) {
            currentList.add(nums[i])
            backtracking(nums, i + 1, currentList)
            currentList.removeAt(currentList.size - 1)
        }
    }
}
